### PR TITLE
nonminimal_bool fix double not

### PR DESCRIPTION
**Pull request by disco07**
fix issue https://github.com/rust-lang/rust-clippy/issues/10836

changelog: Fix [`nonminimal_bool`] false positive when `!!x`, `x` isn't boolean and implements `Not`

### CODE REVIEW

File: clippy_lints/src/booleans.rs [hunk 1]
```diff
@@ -473,6 +496,12 @@ impl<'a, 'tcx> Visitor<'tcx> for NonminimalBoolVisitor<'a, 'tcx> {
                     self.bool_expr(e);
                 },
                 ExprKind::Unary(UnOp::Not, inner) => {
+                    if let ExprKind::Unary(UnOp::Not, ex) = inner.kind {
+                        let ty = self.cx.typeck_results().expr_ty(ex);
+                        if is_impl_not_trait_with_bool_out(self.cx, ty) {
+                            return;
+                        }
```
Comment (samueltardieu) [inline clippy_lints/src/booleans.rs:503]: Out of curiosity, why not just check if `ex` is a `bool`?

**Comment by samueltardieu** [comment]
> changelog: none

There probably should be a changelog there as this is a bugfix to an existing lint.

File: clippy_lints/src/booleans.rs (hunk 1 above)
Comment (disco07) [reply clippy_lints/src/booleans.rs:503]: Thanks you for your feedback. I try this

Comment (disco07) [reply clippy_lints/src/booleans.rs:503]: Are you some method to suggest to me please? When I try `cx.typeck_results().expr_ty(..)` on struct, it returns struct name, not bool type

**Comment by samueltardieu** [comment]
Note that this only seems to fix the particular double not case. Other cases are problematic, such as 

```rust
   let _a = !(!S && !S);
```

when `S` implements `std::ops::Not<Output = bool>`: Clippy will suggest to replace this with `S`.

**Pull request merged**